### PR TITLE
Update "Clone Git Repository" to confirm if and when clone operation is successful

### PR DESCRIPTION
**Pull request by pranavsuryaw**
Signed-off-by: Pranav Suryawanshi <prsuryaw@redhat.com>

### CODE REVIEW

File: tests/Tests/500__jupyterhub/temp.robot
Comment (lugi0) [inline tests/Tests/500__jupyterhub/temp.robot:56]: rename `Get Directory Name From Git Repo URL`

File: tests/Tests/500__jupyterhub/temp.robot
Comment (lugi0) [inline tests/Tests/500__jupyterhub/temp.robot:45]: Rename to `Did Error Occur While Cloning Git Repo`

File: tests/Tests/500__jupyterhub/temp.robot
Comment (pranavsuryaw) [reply tests/Tests/500__jupyterhub/temp.robot]: Done

File: tests/Tests/500__jupyterhub/temp.robot
Comment (pranavsuryaw) [reply tests/Tests/500__jupyterhub/temp.robot]: Done

File: tests/Tests/500__jupyterhub/temp.robot
Comment (pranavsuryaw) [reply tests/Tests/500__jupyterhub/temp.robot]: Done

File: tests/Tests/500__jupyterhub/temp.robot
Comment (pranavsuryaw) [reply tests/Tests/500__jupyterhub/temp.robot]: Done

**Comment by pranavsuryaw** [comment]
https://opendatascience-jenkins-csb-rhods.apps.ocp-c1.prod.psi.redhat.com/job/rhods-ci-pr-test/665/console
Result : passed

**Comment by lugi0** [comment]
@pranavsuryaw Thanks, looks good. Please run the full Sanity job with this PR since you are modifying a keyword a lot of test cases are using

**Review by lugi0 — DISMISSED** [review]
lgtm, needs full sanity run

**Comment by pranavsuryaw** [comment]
https://opendatascience-jenkins-csb-rhods.apps.ocp-c1.prod.psi.redhat.com/job/rhods-ci-pr-test/677/console

**Comment by lugi0** [comment]
@pranavsuryaw there are 52 failures, clearly something went wrong with that run. Please run it again next week on an isolated cluster.

**Comment by pranavsuryaw** [comment]
https://opendatascience-jenkins-csb-rhods.apps.ocp-c1.prod.psi.redhat.com/job/rhods-ci-pr-test/682/console

File: tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot
Comment (jgarciao) [inline tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot:236]: There was already `Clone Git Repository` and `Clone Git Repository And Open` and this RP is adding two new keywords, `Clone Repo` and `Clone Repo and Return Error Message`.  I think it will add more confusion. 

Is it possible just modify `Clone Git Repository`  for the new needs?
- Add a parameter ${delete-existing-clone}  (enabled by default) the will preventively delete the folder if it exists already
- Log error messages

File: tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot
Comment (jgarciao) [inline tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot:240]: Shouldn’t it do this is status <> PASS ?

File: tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot
Comment (jgarciao) [inline tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot:425]: Arguments should not be in capital letters

File: tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot
Comment (jgarciao) [inline tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot:433]: Arguments should not be in capital letters

File: tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot
Comment (jgarciao) [inline tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot:441]: Arguments should not be in capital letters

File: tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot
Comment (jgarciao) [inline tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot:446]: This seems more explanatory
```
Get Git Clone Error Message
[Documentation]    Returns expected error after a git clone operation. Fails if error didn't occur
```

File: tests/Tests/500__jupyterhub/temp.robot
Comment (jgarciao) [inline tests/Tests/500__jupyterhub/temp.robot:1]: Could you rename temp.robot to a more meaningful filename?

File: tests/Tests/500__jupyterhub/temp.robot
Comment (jgarciao) [inline tests/Tests/500__jupyterhub/temp.robot:23]: A more explanatory title is needed

File: tests/Tests/500__jupyterhub/temp.robot
Comment (jgarciao) [inline tests/Tests/500__jupyterhub/temp.robot:25]: Missing tags

File: tests/Tests/500__jupyterhub/temp.robot
Comment (jgarciao) [inline tests/Tests/500__jupyterhub/temp.robot:28]: `Clean Up Server` probably needs to be in `Test Teardown`

**Review by jgarciao — CHANGES_REQUESTED** [review]

File: tests/Tests/500__jupyterhub/temp.robot
Comment (pranavsuryaw) [reply tests/Tests/500__jupyterhub/temp.robot:1]: I have created this file temporarily to test the keyword and this file will be deleted once it gets the verified tag
I have written the same at line no.1 and 2

File: tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot
Comment (pranavsuryaw) [reply tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot]: Done

File: tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot
Comment (pranavsuryaw) [reply tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot]: Done

File: tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot
Comment (pranavsuryaw) [reply tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot]: Done

File: tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot
Comment (pranavsuryaw) [reply tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot]: Done

File: tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot
Comment (pranavsuryaw) [reply tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot:241]: I tried it, it is not working

File: tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot
Comment (pranavsuryaw) [reply tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot:237]: If we just modify the `Git Clone Repository` keyword without having keywords like `Clone Repo` and 'Clone Repo and Return Error Message' then it will not look good.

**Comment by pranavsuryaw** [comment]
@jgarciao I have made the requested changes in that keyword in temp file, if it looks good then i will update in the JupyterLabLauncher.robot file

File: tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot
Comment (bdattoma) [inline tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot:51]: Camel Case

File: tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot
Comment (bdattoma) [inline tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot:252]: could you explain what's the goal here?

If I got right, if the clone process fails, then it deletes the repo dir (_but was the directory created if the clone process failed?_) and then it tries to clone again. Is that right?

File: tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot
Comment (bdattoma) [inline tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot:254]: why 200s?

File: tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot
Comment (bdattoma) [inline tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot:439]: either this one or `Clone Git Repository` should have different names and doc, otherwise it gets confusing

**Review by bdattoma — CHANGES_REQUESTED** [review]

File: tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot
Comment (jgarciao) [inline tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot:246]: @pranavsuryaw  This keyword will fail if the current user running the notebook is not ${TEST_USER.USERNAME}

To fix it I see two options:
- Add  ${admin_username} and ${username} as parameters to this keyword and the keywords using it, like `Clone Git Repository And Open` and `Clone Git Repository And Run`
- (I prefer this option Do not modify the existing clone keywords or add new ones and just add a new keyword to be explicitly called by the tests ). Something like this:
```
Delete Existing Git Clone Folder In User Notebook
    [Documentation]  In case it exists, deletes recursivelly existing git clone folder in user notebook
    ...   (e.g. /opt/app-root/src/repo-folder in ${username}'s notebook PVC.
    ...   Note: this command requires ${admin_username}  to be logged to the cluster (oc login ...)
    ...   and to have the user's notebook pod in Running state (e.g. jupyterhub-nb-ldap-2duser1)
    [Arguments]    ${admin_username}    ${username}    ${repo_url}

    ${repo_folder} =    Get Directory Name From Git Repo URL    ${repo_url}
    Delete Folder In User Notebook
    ...    admin_username=${admin_username}
    ...    username=${username}
    ...    folder=${repo_folder}

```

@lugi0 @bdattoma @pranavsuryaw What do you think?

**Review by jgarciao — CHANGES_REQUESTED** [review]

File: tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot
Comment (bdattoma) [reply tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot:246]: is the goal here to avoid failure in cloning when the same folder alredy exists in jupyterlab? 

If I got it right, I would say the @jgarciao 's solution of calling the kw to delete the folder from the test case may be good and cleaner. Maybe tht kw can be run with "Run Keyword and Ignore Error".

+1 on writing the error msg

Comment (bdattoma) [reply tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot:246]: If I got it right, I would say the @jgarciao 's solution of calling the kw to delete the folder from the test case may still work by running the deletion kw with "Run Keyword and Ignore Error".

Otherwise, I think fetching the user from the URL is a good idea too. Regarding having a variable `CURRENT_USER`, I think it can bring confusion. I would rather prefer using the arguments

+1 on writing the error msg

**Comment by pranavsuryaw** [comment]
I have made changes to the temp.robot file, I will update the keywords in the JupyterLabLauncher.robot file if it is correct.

File: tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot
Comment (lugi0) [inline tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot:457]: Missing documentation and `Private Keyword` tag

File: tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot
Comment (lugi0) [reply tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot:448]: @pranavsuryaw ^

File: tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot
Comment (lugi0) [inline tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot:450]: Missing tag as `Private Keyword`

File: tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot
Comment (lugi0) [reply tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot:53]: @pranavsuryaw ^

File: tests/Tests/500__jupyterhub/temp.robot
Comment (lugi0) [inline tests/Tests/500__jupyterhub/temp.robot:1]: Needs to be removed

**Review by lugi0 — CHANGES_REQUESTED** [review]
@pranavsuryaw still a few changes needed. Test this by running the whole Sanity suite, please

File: tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot
Comment (pranavsuryaw) [reply tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot:450]: Done

File: tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot
Comment (pranavsuryaw) [reply tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot]: Done

**Review by lugi0 — APPROVED** [review]
it looks good to me, please launch a PR test on this and run the whole Sanity suite for it

**Comment by bdattoma** [comment]
> I don't see any failures due to git clones in the sanity run logs, I think this can be merged. @jgarciao @bdattoma can you also take a look?

lgtm. @jgarciao pls have a look since there are some conversations open who might need to close

**Review by bdattoma — APPROVED** [review]

File: tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot
Comment (jgarciao) [reply tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot:450]: Sorry, it should be just `Private`

File: tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot
Comment (jgarciao) [inline tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot:458]: It should be just `Private`.  

Additionally, you could add in the docs something like this:
[Documentation]    (Private Keyword) Clones the github repository and returns the error

**Review by jgarciao — CHANGES_REQUESTED** [review]

File: tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot
Comment (jgarciao) [reply tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot:450]: Thinking again, as the code has been tested and it has 2 approvals this can be changed in another PR

File: tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot
Comment (jgarciao) [reply tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot:458]: Thinking again, as the code has been tested and it has 2 approvals this can be changed in another PR

**Review by jgarciao — APPROVED** [review]
lgtm

**Comment by jgarciao** [comment]
Just in case, I think we should merge this once the RHODS 1.12 RC build smoke job has been run successfully.

 I've temporally added the "do not merge" tag

File: tests/Tests/500__jupyterhub/temp.robot
Comment (jgarciao) [reply tests/Tests/500__jupyterhub/temp.robot]: As this test will be removed before merging the MR, it's not necessary

File: tests/Tests/500__jupyterhub/temp.robot
Comment (jgarciao) [reply tests/Tests/500__jupyterhub/temp.robot]: As this test will be removed before merging the MR, it's not necessary

File: tests/Tests/500__jupyterhub/temp.robot
Comment (jgarciao) [reply tests/Tests/500__jupyterhub/temp.robot]: As this test will be removed before merging the MR, it's not necessary

File: tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot
Comment (jgarciao) [reply tests/Resources/Page/ODH/JupyterHub/JupyterLabLauncher.robot]: The agreement was to obtain the username using @lugi0 proposal

**Comment by lugi0** [comment]
@jgarciao Build has been handed off, should we merge this?

**Pull request merged**
